ramips: Further improvements and fixes of MAC address setup

adschm opened the pull request:
Based on recent device support discussion and fixes on ramips MAC address assignment, this patchset removes the misleading wan_mac default cases and fixes/tidies up several things after that.

Please refer to the individual commit description for further details.

adrianschmutzler commented:
Despite what's addressed in the patches, I left some devices untouched which show strange setup:
onion,omega2 / onion,omega2p :
`mtd-mac-address = <&factory 0x28>;`
`wan_mac=$(mtd_get_mac_binary factory 4)`
`lan_mac=$(mtd_get_mac_binary factory 46)`

_Three_ addresses are used (strange thing no. 1: why are lan AND wan different from &ethernet).
But the device is set up as one-port (strange thing no. 2):
```
		ucidef_add_switch "switch0"
		ucidef_add_switch_attr "switch0" "enable" "false"
		ucidef_set_interface_lan "eth0"
```
Multiple examples can be found for both cases.

adrianschmutzler commented:
Rebased.

adrianschmutzler commented:
@dangowrt 
Hi, I'm currently reviewing the MAC address assignment on ramips target.
You added support for zbt-we1226 some while ago, and set &ethernet to <&factory 0x2e>, with WAN address just calculated by adding +1 to that.
However, there might be another MAC address stored on flash for the device in <&factory 0x28>.

Can you check that? If yes, I would also be interested in which of the two addresses should be "lan" and which "wan" then ...

dangowrt commented:
unfortunately I'm not around that piece of hardware any longer :(

adrianschmutzler commented:
@pepe2k
Hi, I'm currently reviewing the MAC address assignment on ramips target.
You added support for ALFA Network AC1200RM some while ago, and set &ethernet to <&factory 0x28>, with WAN address just calculated by adding +1 to that.
However, there might be another MAC address stored on flash for the device in <&factory 0x2e>.

Can you check that? If yes, I would also be interested in which of the two addresses should be "lan" and which "wan" then ...

adrianschmutzler commented:
@pepe2k Same question for u7628-01-xxx devices.

adrianschmutzler commented:
For later reference: Discussion about newifi devices can be found here: https://github.com/openwrt/openwrt/pull/2075#issuecomment-521985197

adrianschmutzler commented:
Rebased.

adrianschmutzler commented:
@981213 @blocktrron 
I do not think there will be a lot of additional feedback here, so have a look at this if you like.
It contains several fixes/improvements besides changing the default case logic.

adrianschmutzler commented:
@zorun
I'm currently looking into the MAC address assignment of several ramips devices (openwrt/openwrt#2319).
Many of them just have the wan MAC address calculated although it is stored on flash.

As you happen to own a ZBT WE3526, can you check whether you have a MAC address in <&factory 0xe006> and whether it matches your wan (eth0.2) MAC address? If you could provide a hexdump of your factory partition, this would be even better ...

zorun reviewed (commented):
What happens when a device does not match any case? I.e. `wan_mac` / `lan_mac` is not defined?

zorun reviewed (commented):
This patch is really hard to review because you bundle several changes together... It's hard to check what is an actual change and what is just cosmetic or reordering, and whether any device was forgotten/lost in the change.

zorun commented:
> @zorun
> I'm currently looking into the MAC address assignment of several ramips devices (#2319).
> Many of them just have the wan MAC address calculated although it is stored on flash.
> 
> As you happen to own a ZBT WE3526, can you check whether you have a MAC address in <&factory 0xe006> and whether it matches your wan (eth0.2) MAC address? If you could provide a hexdump of your factory partition, this would be even better ...

So, here it is:

- MAC address on the label: `78:A3:51:30:18:F8`
- MAC address on `eth0.1` (LAN): `78:A3:51:30:18:FA` (with OpenWrt `r10865-c3a78955f3`)
- MAC address on `eth0.2` (WAN): `78:A3:51:30:18:FB` (with OpenWrt `r10865-c3a78955f3`)

The data at 0xe000 is `78 a3 51 30 18 fa`, and data at 0xe006 is `78 a3  51 30 18 fb`.

Here are the MAC addresses I spotted in the factory partition:

- 0x0004: `78 a3 51 30  18 f8` (the one from the label)
- 0x8004: `78 a3 51 30  18 f9`
- 0xe000: `78 a3 51 30 18 fa`
- 0xe006: `78 a3  51 30 18 fb`

So, either the label is wrong, or the current OpenWrt location is wrong.

Here is the full hexdump if you find it useful: https://paste.swordarmor.fr/Z3Xg

adrianschmutzler commented:
> What happens when a device does not match any case? I.e. wan_mac / lan_mac is not defined?

@zorun 

If set up at all, wan will be some ethY.2 and lan will be some ethX.1, where in most cases ethX and ethY will be eth0. So, ethX.1 will inherit the MAC address from ethX and ethY.2 will inherit the MAC address from ethY. If you set lan_mac or wan_mac, you _overwrite_ this inherent MAC address.
That's why it makes no sense to set BOTH lan_mac and wan_mac, as then the address from eth0 will be completely unused. Also setting eth0.2 to eth0 address makes no sense here.
Note that you do only set the ethX.Z, not the lan/wan bridge/device directly ...

At the moment (without the patchset), if there is no specific definition for a device in 02_network, the default case will just set eth0 + 1 for it. This might be based on a frequent case, but I think it has several drawbacks, so my first patch here changes the default case to be what one would expect: no change to the MAC addresses if you do not set one.

The other patches then fix and enhance stuff. Note that some of those changes are only possible after the default case has been removed.

> This patch is really hard to review because you bundle several changes together... 

I have thought about that, but I doubt it will be easier with more patches after all. However, I think about giving more details in the commit message for the big patches ...

adrianschmutzler commented:
> MAC address on eth0.2 (WAN): 78:A3:51:30:18:FB (with OpenWrt r10865-c3a78955f3)
>
> 0xe006: 78 a3 51 30 18 fb

Well, that's what I wanted to know. So, the wan address can actually be read from 0xe006 instead of calculating it.

> So, either the label is wrong, or the current OpenWrt location is wrong.

0x0004 and 0x8004 are used for Wifi when reading caldata from 0x0000 and 0x8000, so everything looks correct there to me. Or are you referring to something else?

Thank you very much for the detailed information.

adrianschmutzler commented:
Update:
- Added wan_mac flash location for we3526
- Added extensive descriptions for the first two patches, to make the approach more comprehensible

adrianschmutzler commented:
Added we1326.

adrianschmutzler commented:
Sent the following two patches to the list, as they are simple fixes and independent of the other changes here:
- ramips: remove duplicate case for MAC setup of freestation5
- ramips: fix duplicate network setup for dlink,dir-615-h1

981213 commented:
Hi! I'm a bit skeptical about this: ramips: remove default case for MAC address assignment

> This is not desirable, as many device supporters will just not
> care or know about this definition, so another MAC address will be
> introduced by accident. In some cases the wan_mac is assigned
> although it is not needed, in other cases even addresses not
> dedicated to the device will be used

Those developers who are aware of this problem will add correct definition, and those who don't care will just add their device into the largest case anyway. In the end we just bloat 02_network a bit and still have to explicitly ask PR/patch authors to check mac address assignments when they use wan_mac = lan_mac + 1

BTW I'd be happy to merge the rest of this PR if you could move this commit down to the last one :)

adrianschmutzler commented:
@981213 Thanks for response.

> In the end we just bloat 02_network a bit and still have to explicitly ask PR/patch authors to check mac address assignments when they use wan_mac = lan_mac + 1

Well, that's part of what I want to achieve, since I have the feeling that almost nobody really cares about the MAC assignment. In this situation, I prefer that the event leading to another MAC address occupied requires more work than the event where nothing is changed.

But despite that, note that especially patch 2 here is based on having the default case removed:
With the default changed, I can remove entire blocks that do not set anything in patch 2. If the "old" default case remained, I would have to introduce a "do nothing" case into 02_network, just to prevent devices falling into the default and having wan set to lan+1 although that's not correct.
While this "do nothing" case might be considerably smaller that the lan+1 case, it still feels wrong to me to have an "empty" case where the default case does something different.
As already mentioned, this will have the counter-intuitive effect that removing a case will not result in having no MAC address changes, but in having just different address changes (lan+1).

So I'm still deeply convinced that having the default case swapped is the proper solution here.

> BTW I'd be happy to merge the rest of this PR if you could move this commit down to the last one :)

Since this won't hurt too much, I will try to do that during the day. This will then introduce the "do nothing" case which will be removed again in the final patch. Maybe by that it will also become a little more obvious what my desire is in this context.

adrianschmutzler commented:
@981213 I have changed patch order as requested (including the necessary adjustments).

981213 commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -515,9 +515,13 @@ ramips_setup_macs()
 		;;
+	adslr,g7)
+		wan_mac=$(mtd_get_mac_binary factory 0x4)
+		;;
```
Considering that this router has a label_mac stored at 0xe00c, it's unlikely that vendor decided to use 0xe00c for lan_mac while also using mac for wifi card as wan_mac.
I'd prefer to keep this router to use lan_mac+1 as wan_mac and will patch this away when applying.

adrianschmutzler replied:
Okay, I accept that reasoning.

adrianschmutzler replied:
BTW: If you want to get rid of the lengthy explanation, it's okay if you just use my Signed-off-by for this change.

981213 commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -535,9 +538,12 @@ ramips_setup_macs()
 	phicomm,k2p|\
-	planex,vr500)
-		lan_mac=$(mtd_get_mac_binary factory 0xe000)
+	planex,vr500|\
```
Have you confirmed that this router has a wan_mac stored at 0xe006?

adrianschmutzler replied:
No, I haven't. This patch is only about moving code without changing the behavior I found in this file.
So, I do not know whether the wan_mac is right or wrong, but I do not have to since I'm not changing it.

Changes based on actual tests/confirmations are only implemented in "ramips: initialize MAC addresses from flash where possible" and Newifi Y1(s) patch

981213 replied:
Oh. I misread the patch a bit :)

adrianschmutzler replied:
Yes, for changes like this the git diff is quite misleading sometimes.

981213 reviewed (commented):
Partially merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Just one more question before pushing:

adrianschmutzler commented:
Thanks for merge.
Rebased and updated the rest.

adrianschmutzler commented:
@981213 I've backported the most important fixes and sent them to the list...

adrianschmutzler commented:
Rebased.

adrianschmutzler commented:
Change for youhua,wr1200js has been sent to list as separate patch.

adrianschmutzler commented:
Rebased and adjusted due to unielec rename.

adschm committed:
ramips: remove default case for MAC address assignment

So far, MAC address assignment in ramips has contained a default
case, which defined wan_mac = eth0 + 1 for _every_ device not
having an explicit case there.

This is not desirable, as many device supporters will just not
care or know about this definition, so another MAC address will be
introduced by accident. In some cases the wan_mac is assigned
although it is not needed, in other cases even addresses not
dedicated to the device will be used (e.g. wan_mac actually is
eth0 - 1, but during support nobody cared, so eth0 + 1 is used now,
which might actually belong to another device ...).

Thus, in this PR the former default case is converted to an
explicit case. This one comprises all devices not being accounted
for by other cases, reduced by those not having wan at all.
The big number of entries for this node might be another indication
that many of them wouldn't actually be there if there hadn't been
default wan_mac setup.

In exchange, the current "do nothing" case can be removed, as it
will be the new default case.

The devices being put in the newly created explicit case were
determined as follows:

1. Create a list of all devices based on the DTS files.

2. Remove all devices already having an explicit entry setting
   their address.

3. Remove all devices that only have lan set up in the first part
   of 02_network:

mt7620:
   - alfa-network,tube-e4g
   - asus,rp-n53
   - buffalo,wmr-300
   - comfast,cf-wr800n
   - edimax,ew-7476rpc
   - edimax,ew-7478ac
   - elecom,wrh-300cr
   - hnet,c108
   - kimax,u25awf-h1
   - kimax,u35wf
   - kingston,mlw221
   - kingston,mlwg2
   - microduino,microwrt
   - netgear,ex2700
   - netgear,ex3700
   - netgear,wn3000rp-v3
   - planex,cs-qr10
   - planex,mzk-ex300np
   - planex,mzk-ex750np
   - ravpower,wd03
   - sercomm,na930
   - yukai,bocco
   - zbtlink,zbt-cpe102
   - zbtlink,zbt-we1026-5g-16m
   - zte,q7

mt7621:
   - gnubee,gb-pc1
   - gnubee,gb-pc2
   - linksys,re6500
   - mikrotik,rbm11g
   - netgear,ex6150
   - thunder,timecloud
   - tplink,re350-v1
   - tplink,re650-v1

mt76x8:
   - alfa-network,awusfree1
   - d-team,pbr-d1
   - glinet,vixmini
   - vocore,vocore2-lite
   - tama,w06
   - tplink,tl-mr3020-v3
   - tplink,tl-wa801nd-v5
   - tplink,tl-wr802n-v4
   - tplink,tl-wr902ac-v3
   - vocore,vocore2
   - widora,neo-16m
   - widora,neo-32m

rt288x:
   - buffalo,wli-tx4-ag300n
   - dlink,dap-1522-a1

rt305x:
   - allnet,all0256n-4m
   - allnet,all0256n-8m
   - allnet,all5002
   - allnet,all5003
   - alphanetworks,asl26555-16m
   - alphanetworks,asl26555-8m
   - asus,wl-330n
   - aximcom,mr-102n
   - dlink,dcs-930
   - easyacc,wizard-8800
   - hame,mpr-a2
   - hootoo,ht-tm02
   - huawei,d105
   - intenso,memory2move
   - planex,mzk-dp150n
   - rt305x dlink,dcs-930l-b1
   - sparklan,wcr-150gn
   - tenda,3g150b
   - tenda,3g300m
   - tenda,w150m
   - trendnet,tew-638apb-v2
   - unbranded,a5-v11
   - vocore,vocore-16m
   - vocore,vocore-8m
   - wansview,ncs601w
   - zorlik,zl5900v2

rt3883:
   - loewe,wmdr-143n
   - omnima,hpm

4. Put the remaining devices in the new case.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
@981213 I personally still think that the wan MAC assignment patch here would be an improvement. 
I've rebased the patch onto the base-files split, so it is now distributed on the 6 new 02_network files.

Since you have been skeptical about it and justified your view, I want to give you the chance to veto on merging this patch before I push it.

adrianschmutzler commented:
I will merge this today.
